### PR TITLE
[REFACTORING] making code more compact

### PR DESCRIPTION
**Pull request by dennisseah**
1. move command description to a json file
2. create a command builder class
3. make `validateRequiredArguments` function non `async`
4. remove unused code in test file
5. create helper function for checking missing values
6. remove `any` type with proper type
7. add types to function
8. use process.exit(0) and process.exit(1) for terminating command 

Related to https://github.com/microsoft/bedrock/issues/885

### CODE REVIEW

**Comment by andrebriggs** [comment]
Thanks @dennisseah for the pull request. Currently the checks for your PR area failing because there is a drop in overall branch line coverage that is more than 1%.

> The code coverage value (58.4939%, 668 branches) is lower than the value of the previous build including variance! The previous build had a coverage value of 59.7964% (705 branches).

## Your PR at https://dev.azure.com/epicstuff/bedrock/_build/results?buildId=9807&view=codecoverage-tab
image

## Last commit to master at https://dev.azure.com/epicstuff/bedrock/_build/results?buildId=9797&view=codecoverage-tab
image

The links are a good starting point at looking at the coverage differences.

**Comment by mtarng** [comment]
Seems like this could also be combined with this ticket https://github.com/microsoft/bedrock/issues/741
We are looking to standardize flag names and shorthand flags across all commands, though we don't have a design proposed yet. I like what you've done with the decorator.json file, but we'll probably want a central library for all commands to share/overlap. Command builder is great too.

Thoughts? @sarath-p @evanlouie @bnookala

**Comment by dennisseah** [comment]
> Seems like this could also be combined with this ticket [microsoft/bedrock#741](https://github.com/microsoft/bedrock/issues/741)
> We are looking to standardize flag names and shorthand flags across all commands, though we don't have a design proposed yet. I like what you've done with the decorator.json file, but we'll probably want a central library for all commands to share/overlap. Command builder is great too.
> 
> Thoughts? @sarath-p @evanlouie @bnookala

I was combing thru the command and found that sometimes, we use --organization-name and other times, we use --org-name. Having in json, allow us to write code to scan for consistency. BTW, I found a bug, we have --build-script (like here, https://github.com/CatalystCode/spk/blob/getting-started-docs/src/commands/project/pipeline.ts#L45) and then we get the value via buildScriptUrl instead of buildScript (here https://github.com/CatalystCode/spk/blob/getting-started-docs/src/commands/project/pipeline.ts#L60)

**Comment by mtarng** [comment]
yup. If you look at the ticket, there's a spreadsheet of all the current commands (https://microsoft.sharepoint.com/:x:/t/DevCrews-Episerver/EUA0JeUOgppHq_YY6sAkD68BYa3aHX1idc8cwOLgRT0gYg?e=sUo0jx)

Definitely want to get parity with all commands

**Review by andrebriggs — APPROVED** [review]

**Comment by evanlouie** [comment]
was their a particular reason we chose to use json files for the decorators specs as opposed to exporting a JS object? I think some structural integrity would be added if we did that and had a shared interface for the decorators to export.

something like

```ts
interface ICommandDecorator {
  command: string;
  alias?: string;
  description: string;
  options?: {
    [flag: string]: {
      arg?: string;
      description: string;
      required: boolean;
    };
  };
}
```

File: src/commands/project/create-variable-group.ts [hunk 1]
```diff
@@ -3,105 +3,112 @@ import commander from "commander";
...
+  } else {
+    try {
+      const { azure_devops } = Config();
+
+      const {
+        registryName,
+        servicePrincipalId,
+        servicePrincipalPassword,
+        tenant,
+        hldRepoUrl = azure_devops && azure_devops.hld_repository,
```
Comment (evanlouie) [inline src/commands/project/create-variable-group.ts]: we should avoid using `&&` for de-structuring as we are now on TS 3.7.
Should use `hldRepoUrl = azure_devops?.hld_repository` instead

File: src/commands/project/create-variable-group.ts [hunk 2]
```diff
@@ -3,105 +3,112 @@ import commander from "commander";
...
+        orgName = azure_devops && azure_devops.org,
```
Comment (evanlouie) [inline src/commands/project/create-variable-group.ts]: Use `?.` operator instead of `&&`

File: src/commands/project/create-variable-group.ts [hunk 3]
```diff
@@ -3,105 +3,112 @@ import commander from "commander";
...
+        personalAccessToken = azure_devops && azure_devops.access_token,
```
Comment (evanlouie) [inline src/commands/project/create-variable-group.ts]: Use `?.` operator instead of &&

File: src/commands/project/create-variable-group.ts [hunk 4]
```diff
@@ -3,105 +3,112 @@ import commander from "commander";
...
+        project = azure_devops && azure_devops.project
```
Comment (evanlouie) [inline src/commands/project/create-variable-group.ts]: Use `?.` operator instead of &&

**Comment by dennisseah** [comment]
> was their a particular reason we chose to use json files for the decorators specs as opposed to exporting a JS object? I think some structural integrity would be added if we did that and had a shared interface for the decorators to export.
> 
> something like
> 
> ```ts
> interface ICommandDecorator {
>   command: string;
>   alias?: string;
>   description: string;
>   options?: {
>     [flag: string]: {
>       arg?: string;
>       description: string;
>       required: boolean;
>     };
>   };
> }
> ```

Thanks for your comment, @evan

1. JSON is lightweight. We can parse it and generate documentation for the command line if we wish to (later in time)
2. during JSON import and calling the build function in CommandBuilder, the type checking is done. that’s i got an error if my json is incorrect (does not follow the interface)

File: src/lib/commandBuilder.test.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        },
+        optionC: {
+          arg: "-c, --optionC <optionC>",
+          description: "description for optionC",
+          required: false
+        }
+      }
+    };
+
+    const options: ICommandOption[] = Object.getOwnPropertyNames(
```
Comment (evanlouie) [inline src/lib/commandBuilder.test.ts]: should use `.keys` instead of `.getOwnPropertyNames`

File: src/lib/commandBuilder.test.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        },
+        optionC: {
+          arg: "-c --optionC <optionC>",
+          description: "description for optionC",
+          required: true
+        }
+      }
+    };
+
+    const options: ICommandOption[] = Object.getOwnPropertyNames(
```
Comment (evanlouie) [inline src/lib/commandBuilder.test.ts]: `.keys` instead of `.getOwnPropertyNames`

File: src/lib/commandBuilder.ts [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+  command: commander.Command,
+  decorator: ICommandBuildElements
+): commander.Command => {
+  const cmd = command
+    .command(decorator.command)
+    .alias(decorator.alias)
+    .description(decorator.description);
+
+  const options = decorator.options;
+  Object.getOwnPropertyNames(options).forEach(name => {
```
Comment (evanlouie) [inline src/lib/commandBuilder.ts]: `.keys` instead of `.getOwnPropertyNames`

File: src/lib/commandBuilder.ts [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+ * @param decorator Descriptor for command building
+ * @param values Values to be inspected.
+ * @return error messages.
+ */
+export const validateForRequiredValues = (
+  decorator: ICommandBuildElements,
+  values: { [key: string]: string | undefined }
+): string[] => {
+  // gather the required variable names from the decorator options object
+  const requireds = Object.getOwnPropertyNames(decorator.options).filter(
```
Comment (evanlouie) [inline src/lib/commandBuilder.ts]: `.keys` instead of `.getOwnPropertyNames`

File: src/lib/commandBuilder.ts [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+ * Command Option
+ */
+export interface ICommandOption {
+  arg: string;
+  description: string;
+  required: boolean;
+}
+
+/**
+ * Map of key (stirng) to its command option
```
Comment (evanlouie) [inline src/lib/commandBuilder.ts]: typo (strirng => string)

**Review by evanlouie — CHANGES_REQUESTED** [review]
some initial quick finds

File: src/lib/validator.ts [hunk 10]
```diff
@@ -0,0 +1,29 @@
+/**
+ * Values to be validated
+ */
+export interface IValidationValue {
+  value: undefined | null | string;
+  error: string;
+}
+
+/**
+ * Returns true of val is undefined, null or empty string.
```
Comment (evanlouie) [inline src/lib/validator.ts]: I believe this doc is incorrect. Should be `Returns true if val is not undefined, not null, and not an empty string.`

File: src/lib/validator.ts [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ * @param val Value to inspect
+ */
+export const hasValue = (val: undefined | null | string): boolean => {
```
Comment (evanlouie) [inline src/lib/validator.ts]: function should be changed to a type-guard

```ts
export const hasValue = (val: undefined | null | string): val is string => {
  return val !== undefined && val !== null && val !== "";
};
```

**Review by evanlouie — CHANGES_REQUESTED** [review]

File: src/commands/project/create-variable-group.ts (hunk 1 above)
Comment (dennisseah) [reply src/commands/project/create-variable-group.ts]: ok. thinking not to change the initial code too much

File: src/commands/project/create-variable-group.ts (hunk 2 above)
Comment (dennisseah) [reply src/commands/project/create-variable-group.ts]: ok. thinking not to change the initial code too much

File: src/commands/project/create-variable-group.ts (hunk 3 above)
Comment (dennisseah) [reply src/commands/project/create-variable-group.ts]: ok. thinking not to change the initial code too much

File: src/commands/project/create-variable-group.ts (hunk 4 above)
Comment (dennisseah) [reply src/commands/project/create-variable-group.ts]: ok. thinking not to change the initial code too much

File: src/lib/commandBuilder.test.ts (hunk 5 above)
Comment (dennisseah) [reply src/lib/commandBuilder.test.ts]: refactor the code and this is not needed now. thanks

File: src/lib/commandBuilder.test.ts (hunk 6 above)
Comment (dennisseah) [reply src/lib/commandBuilder.test.ts]: refactor the code and this is not needed now. thanks

File: src/lib/commandBuilder.ts (hunk 7 above)
Comment (dennisseah) [reply src/lib/commandBuilder.ts]: refactor the code and this is not needed now. thanks

File: src/lib/commandBuilder.ts (hunk 8 above)
Comment (dennisseah) [reply src/lib/commandBuilder.ts]: refactor the code and this is not needed now. thanks

File: src/lib/commandBuilder.ts (hunk 9 above)
Comment (dennisseah) [reply src/lib/commandBuilder.ts]: accept thanks

File: src/lib/validator.ts (hunk 10 above)
Comment (dennisseah) [reply src/lib/validator.ts]: accept

File: src/commands/infra/scaffold.ts [hunk 12]
```diff
@@ -5,9 +5,86 @@ import yaml from "js-yaml";
...
+      removeTemplateFiles(opts.name);
+      exitFn(0);
+    } catch (err) {
+      logger.error("Error occurred while generating scaffold");
+      logger.error(err);
+      exitFn(1);
+    }
+  }
+};
+
```
Comment (yradsmikham) [inline src/commands/infra/scaffold.ts]: Somewhere in this refactoring is breaking the functionality of `spk infra scaffold`. After testing the code, I am unable to generate a ` definition.yaml` file. I tried debugging, but am unable to find the root cause. Will need to dive into it a little deeper...

But you if you run `spk infra scaffold` against the spk master branch, and one against your forked branch, you'll see that in the refactored code, scaffolding fails to generate the definition and remove the templated terraform files.

SPK master branch:
Screen Shot 2020-01-22 at 8 57 30 AM

Your SPK master branch:
Screen Shot 2020-01-22 at 8 58 33 AM

Are you able to reproduce the same behavior on your side?

**Review by yradsmikham — CHANGES_REQUESTED** [review]

File: src/commands/infra/scaffold.ts (hunk 12 above)
Comment (dennisseah) [reply src/commands/infra/scaffold.ts]: chat with Yvonne and figure out the problem.

**Comment by yradsmikham** [comment]
LGTM

**Review by yradsmikham — APPROVED** [review]

**Review by evanlouie — APPROVED** [review]

**Pull request merged**